File: clients/ts/src/state_machine/caught-up.ts
import type {
  ProtocolQueueAdapter,
  ProtocolSyncCycleResult,
  ProtocolSyncState,
} from '../sync-loop.js';
import { systemCaughtUpNow } from '../runtime/caught-up-clock.js';

const CANONICAL_CHECKPOINT = /^(?:0|[1-9]\d*)$/;

/**
 * An authoritative source position representing all changes committed before
 * the checkpoint request completed.
 *
 * `generation` is optional in protocol-v1 so existing full-replication servers
 * can adopt the barrier without inventing scope semantics. Selective-sync
 * scopes must bind it once scope identity lands (DEN-139).
 */
export interface AuthoritativeCheckpointTarget {
  protocolVersion: 1;
  checkpoint: string;
  generation?: string;
}

/** Optional transport capability; deliberately separate from ProtocolTransport. */
export interface AuthoritativeCheckpointRequester {
  requestCheckpoint(signal: AbortSignal): Promise<AuthoritativeCheckpointTarget>;
}

/** Structural sync-loop surface used by the barrier. */
export interface CaughtUpSyncLoop {
  readonly state?: Readonly<ProtocolSyncState>;
  syncNow(): Promise<ProtocolSyncCycleResult>;
}

export interface AwaitCaughtUpOptions {
  /** Bound the caller's wait. The shared sync cycle itself is never aborted. */
  timeoutMs?: number;
  /** Cancels this waiter only; it never cancels another waiter/background sync. */
  signal?: AbortSignal;
  /** Poll delay when a successful cycle made no checkpoint progress. */
  pollIntervalMs?: number;
  /** Override connectivity detection for non-browser hosts/tests. */
  isOnline?: () => boolean;
  /** Expected checkpoint namespace/reset/scope generation, when one exists. */
  expectedGeneration?: string;
  /** Injectable wall clock for elapsed-time reporting/tests. */
  now?: () => number;
}

export type CaughtUpBarrierErrorCode =
  | 'CAUGHT_UP_INVALID_TARGET'
  | 'CAUGHT_UP_INVALID_LOCAL_CHECKPOINT'
  | 'CAUGHT_UP_INVALIDATED'
  | 'CAUGHT_UP_TIMEOUT'
  | 'CAUGHT_UP_CANCELLED'
  | 'CAUGHT_UP_OFFLINE';

export interface CaughtUpBarrierFailure {
  kind: 'barrier';
  code: CaughtUpBarrierErrorCode;
  message: string;
  targetCheckpoint?: string;
  localCheckpoint?: string;
}

export interface CaughtUpSyncFailure {
  kind: 'sync';
  message: string;
  cause: unknown;
  targetCheckpoint: string;
  localCheckpoint: string;
}

export interface CaughtUpRequestFailure {
  kind: 'request';
  message: string;
  cause: unknown;
}

export type CaughtUpFailure =
  | CaughtUpBarrierFailure
  | CaughtUpSyncFailure
  | CaughtUpRequestFailure;

export interface CaughtUpResult {
  targetCheckpoint: string;
  checkpoint: string;
  generation?: string;
  elapsedMs: number;
  cycles: number;
  alreadyCaughtUp: boolean;
}

export type CaughtUpOutcome<T = CaughtUpResult> =
  | { ok: true; value: T }
  | { ok: false; error: CaughtUpFailure };

const success = <T>(value: T): CaughtUpOutcome<T> => ({ ok: true, value });
const failure = <T = never>(error: CaughtUpFailure): CaughtUpOutcome<T> => ({
  ok: false,
  error,
});

function barrierFailure(
  code: CaughtUpBarrierErrorCode,
  message: string,
  targetCheckpoint?: string,
  localCheckpoint?: string,
): CaughtUpBarrierFailure {
  return {
    kind: 'barrier',
    code,
    message,
    ...(targetCheckpoint === undefined ? {} : { targetCheckpoint }),
    ...(localCheckpoint === undefined ? {} : { localCheckpoint }),
  };
}

function checkpointFailure(
  checkpoint: string,
  local: boolean,
): CaughtUpBarrierFailure | null {
  if (CANONICAL_CHECKPOINT.test(checkpoint)) return null;
  return barrierFailure(
    local ? 'CAUGHT_UP_INVALID_LOCAL_CHECKPOINT' : 'CAUGHT_UP_INVALID_TARGET',
    `${local ? 'local' : 'target'} checkpoint must be a canonical unsigned decimal string`,
    local ? undefined : checkpoint,
    local ? checkpoint : undefined,
  );
}

export function checkpointReached(
  local: string,
  target: string,
): CaughtUpOutcome<boolean> {
  const localError = checkpointFailure(local, true);
  if (localError) return failure(localError);
  const targetError = checkpointFailure(target, false);
  if (targetError) return failure(targetError);
  return success(BigInt(local) >= BigInt(target));
}

export function validateAuthoritativeCheckpointTarget(
  target: AuthoritativeCheckpointTarget,
  expectedGeneration?: string,
): CaughtUpBarrierFailure | null {
  if (target.protocolVersion !== 1) {
    return barrierFailure(
      'CAUGHT_UP_INVALID_TARGET',
      'authoritative checkpoint target has unsupported protocolVersion',
      target.checkpoint,
    );
  }
  const checkpointError = checkpointFailure(target.checkpoint, false);
  if (checkpointError) return checkpointError;
  if (target.generation !== undefined && target.generation.length === 0) {
    return barrierFailure(
      'CAUGHT_UP_INVALID_TARGET',
      'authoritative checkpoint generation must be non-empty when present',
      target.checkpoint,
    );
  }
  if (
    expectedGeneration !== undefined &&
    target.generation !== expectedGeneration
  ) {
    return barrierFailure(
      'CAUGHT_UP_INVALIDATED',
      `checkpoint generation ${JSON.stringify(target.generation ?? null)} does not match expected generation ${JSON.stringify(expectedGeneration)}`,
      target.checkpoint,
    );
  }
  return null;
}

function normalizedNonNegative(
  value: number | undefined,
  fallback: number,
): number {
  if (value === undefined) return fallback;
  if (!Number.isFinite(value)) return fallback;
  return Math.max(0, value);
}

type WaitResult<T> =
  | { kind: 'value'; value: T }
  | { kind: 'operation-error'; cause: unknown }
  | { kind: 'timeout' }
  | { kind: 'cancelled' };

function waitForCaller<T>(
  operation: Promise<T>,
  signal: AbortSignal | undefined,
  timeoutMs: number,
): Promise<WaitResult<T>> {
  const operationResult = operation.then<WaitResult<T>>(
    (value) => ({ kind: 'value', value }),
    (cause: unknown) => ({ kind: 'operation-error', cause }),
  );
  const timeoutResult = new Promise<WaitResult<T>>((resolve) => {
    setTimeout(() => resolve({ kind: 'timeout' }), timeoutMs);
  });
  if (!signal) return Promise.race([operationResult, timeoutResult]);
  if (signal.aborted) return Promise.resolve({ kind: 'cancelled' });
  const cancellationResult = new Promise<WaitResult<T>>((resolve) => {
    signal.addEventListener(
      'abort',
      () => resolve({ kind: 'cancelled' }),
      { once: true },
    );
  });
  return Promise.race([operationResult, timeoutResult, cancellationResult]);
}

function terminalWaitOutcome<T>(
  result: WaitResult<T>,
  target: string,
  local: string,
  syncFailureMessage: string,
): CaughtUpOutcome<T> | null {
  if (result.kind === 'value') return success(result.value);
  if (result.kind === 'operation-error') {
    return failure({
      kind: 'sync',
      message: syncFailureMessage,
      cause: result.cause,
      targetCheckpoint: target,
      localCheckpoint: local,
    });
  }
  if (result.kind === 'cancelled') {
    return failure(
      barrierFailure(
        'CAUGHT_UP_CANCELLED',
        'caught-up wait was cancelled',
        target,
        local,
      ),
    );
  }
  return failure(
    barrierFailure(
      'CAUGHT_UP_TIMEOUT',
      'caught-up wait timed out before the durable checkpoint reached the target',
      target,
      local,
    ),
  );
}

/**
 * Drive the ordinary single-flight sync loop until its *durable* checkpoint is
 * at or beyond `target.checkpoint`.
 */
export async function awaitCaughtUp(
  loop: CaughtUpSyncLoop,
  queue: Pick<ProtocolQueueAdapter, 'pullCheckpoint'>,
  target: AuthoritativeCheckpointTarget,
  options: AwaitCaughtUpOptions = {},
): Promise<CaughtUpOutcome> {
  const targetError = validateAuthoritativeCheckpointTarget(
    target,
    options.expectedGeneration,
  );
  if (targetError) return failure(targetError);
  const now = options.now ?? systemCaughtUpNow;
  const timeoutMs = normalizedNonNegative(options.timeoutMs, 30_000);
  const pollIntervalMs = normalizedNonNegative(options.pollIntervalMs, 50);
  const isOnline =
    options.isOnline ??
    (() => typeof navigator === 'undefined' || navigator.onLine !== false);
  const startedAt = now();
  const deadline = startedAt + timeoutMs;
  let checkpoint = await queue.pullCheckpoint();
  const initialCheckpointError = checkpointFailure(checkpoint, true);
  if (initialCheckpointError) return failure(initialCheckpointError);
  const initialReached = checkpointReached(checkpoint, target.checkpoint);
  if (!initialReached.ok) return initialReached;
  if (initialReached.value) {
    return success({
      targetCheckpoint: target.checkpoint,
      checkpoint,
      generation: target.generation,
      elapsedMs: Math.max(0, now() - startedAt),
      cycles: 0,
      alreadyCaughtUp: true,
    });
  }

  let cycles = 0;
  for (;;) {
    if (options.signal?.aborted) {
      return failure(
        barrierFailure(
          'CAUGHT_UP_CANCELLED',
          'caught-up wait was cancelled',
          target.checkpoint,
          checkpoint,
        ),
      );
    }
    if (!isOnline() || loop.state?.status === 'offline') {
      return failure(
        barrierFailure(
          'CAUGHT_UP_OFFLINE',
          'cannot establish authoritative freshness while offline',
          target.checkpoint,
          checkpoint,
        ),
      );
    }
    const remaining = deadline - now();
    if (remaining <= 0) {
      return failure(
        barrierFailure(
          'CAUGHT_UP_TIMEOUT',
          'caught-up wait timed out before the durable checkpoint reached the target',
          target.checkpoint,
          checkpoint,
        ),
      );
    }

    const before = checkpoint;
    const waited = await waitForCaller(
      loop.syncNow(),
      options.signal,
      remaining,
    );
    const waitOutcome = terminalWaitOutcome(
      waited,
      target.checkpoint,
      checkpoint,
      'protocol sync failed before the caught-up target was reached',
    );
    if (!waitOutcome?.ok) return waitOutcome ?? failure(barrierFailure('CAUGHT_UP_TIMEOUT', 'caught-up wait did not complete', target.checkpoint, checkpoint));
    cycles += 1;

    checkpoint = await queue.pullCheckpoint();
    const checkpointError = checkpointFailure(checkpoint, true);
    if (checkpointError) return failure(checkpointError);
    const reached = checkpointReached(checkpoint, target.checkpoint);
    if (!reached.ok) return reached;
    if (reached.value) {
      return success({
        targetCheckpoint: target.checkpoint,
        checkpoint,
        generation: target.generation,
        elapsedMs: Math.max(0, now() - startedAt),
        cycles,
        alreadyCaughtUp: false,
      });
    }

    if (checkpoint === before && pollIntervalMs > 0) {
      const afterCycleRemaining = deadline - now();
      if (afterCycleRemaining <= 0) {
        return failure(
          barrierFailure(
            'CAUGHT_UP_TIMEOUT',
            'caught-up wait timed out before the durable checkpoint reached the target',
            target.checkpoint,
            checkpoint,
          ),
        );
      }
      const delayed = await waitForCaller(
        new Promise<void>((resolve) =>
          setTimeout(resolve, Math.min(pollIntervalMs, afterCycleRemaining)),
        ),
        options.signal,
        afterCycleRemaining,
      );
      const delayOutcome = terminalWaitOutcome(
        delayed,
        target.checkpoint,
        checkpoint,
        'caught-up poll delay failed',
      );
      if (!delayOutcome?.ok) {
        return delayOutcome ?? failure(barrierFailure('CAUGHT_UP_TIMEOUT', 'caught-up wait did not complete', target.checkpoint, checkpoint));
      }
    }
  }
}

/** Ask the server for a source-now checkpoint, then prove local durability. */
export async function requestAndAwaitCaughtUp(
  loop: CaughtUpSyncLoop,
  queue: Pick<ProtocolQueueAdapter, 'pullCheckpoint'>,
  requester: AuthoritativeCheckpointRequester,
  options: AwaitCaughtUpOptions = {},
): Promise<CaughtUpOutcome> {
  const now = options.now ?? systemCaughtUpNow;
  const timeoutMs = normalizedNonNegative(options.timeoutMs, 30_000);
  const startedAt = now();
  const requestController = new AbortController();
  const forwardAbort = () => requestController.abort();
  options.signal?.addEventListener('abort', forwardAbort, { once: true });
  const requested = await waitForCaller(
    requester.requestCheckpoint(requestController.signal),
    options.signal,
    timeoutMs,
  );
  options.signal?.removeEventListener('abort', forwardAbort);
  if (requested.kind === 'operation-error') {
    return failure({
      kind: 'request',
      message: 'authoritative checkpoint request failed',
      cause: requested.cause,
    });
  }
  if (requested.kind === 'cancelled') {
    requestController.abort();
    return failure(
      barrierFailure('CAUGHT_UP_CANCELLED', 'caught-up wait was cancelled'),
    );
  }
  if (requested.kind === 'timeout') {
    requestController.abort();
    return failure(
      barrierFailure(
        'CAUGHT_UP_TIMEOUT',
        'caught-up wait timed out while requesting the authoritative checkpoint',
      ),
    );
  }
  const elapsed = Math.max(0, now() - startedAt);
  return awaitCaughtUp(loop, queue, requested.value, {
    ...options,
    timeoutMs: Math.max(0, timeoutMs - elapsed),
    now,
  });
}

File: clients/dart/lib/src/state_machine/caught_up.dart
import 'dart:async';

import '../protocol_sync_loop.dart';

final RegExp _canonicalCheckpoint = RegExp(r'^(?:0|[1-9]\d*)$');

class AuthoritativeCheckpointTarget {
  final int protocolVersion;
  final String checkpoint;
  final String? generation;

  const AuthoritativeCheckpointTarget({
    this.protocolVersion = 1,
    required this.checkpoint,
    this.generation,
  });

  factory AuthoritativeCheckpointTarget.fromJson(Map<String, dynamic> json) {
    return AuthoritativeCheckpointTarget(
      protocolVersion: json['protocolVersion'] as int? ?? 0,
      checkpoint: json['checkpoint'] as String? ?? '',
      generation: json['generation'] as String?,
    );
  }

  Map<String, dynamic> toJson() => <String, dynamic>{
    'protocolVersion': protocolVersion,
    'checkpoint': checkpoint,
    if (generation != null) 'generation': generation,
  };
}

abstract interface class AuthoritativeCheckpointRequester {
  Future<AuthoritativeCheckpointTarget> requestCheckpoint(
    CaughtUpCancellationToken cancellation,
  );
}

class CaughtUpCancellationToken {
  bool _cancelled = false;
  final Completer<void> _cancelledCompleter = Completer<void>();

  bool get isCancelled => _cancelled;
  Future<void> get whenCancelled => _cancelledCompleter.future;

  void cancel() {
    if (_cancelled) return;
    _cancelled = true;
    _cancelledCompleter.complete();
  }
}

enum CaughtUpBarrierErrorCode {
  invalidTarget,
  invalidLocalCheckpoint,
  invalidated,
  timeout,
  cancelled,
  offline,
}

enum CaughtUpFailureKind { barrier, sync, request }

class CaughtUpFailure {
  final CaughtUpFailureKind kind;
  final CaughtUpBarrierErrorCode? code;
  final String message;
  final String? targetCheckpoint;
  final String? localCheckpoint;
  final Object? cause;

  const CaughtUpFailure({
    required this.kind,
    required this.message,
    this.code,
    this.targetCheckpoint,
    this.localCheckpoint,
    this.cause,
  });
}

class CaughtUpResult {
  final String targetCheckpoint;
  final String checkpoint;
  final String? generation;
  final Duration elapsed;
  final int cycles;
  final bool alreadyCaughtUp;

  const CaughtUpResult({
    required this.targetCheckpoint,
    required this.checkpoint,
    required this.generation,
    required this.elapsed,
    required this.cycles,
    required this.alreadyCaughtUp,
  });
}

sealed class CaughtUpOutcome<T> {
  const CaughtUpOutcome();
}

final class CaughtUpSuccess<T> extends CaughtUpOutcome<T> {
  final T value;
  const CaughtUpSuccess(this.value);
}

final class CaughtUpFailed<T> extends CaughtUpOutcome<T> {
  final CaughtUpFailure error;
  const CaughtUpFailed(this.error);
}

CaughtUpFailure _barrierFailure(
  CaughtUpBarrierErrorCode code,
  String message, {
  String? target,
  String? local,
}) {
  return CaughtUpFailure(
    kind: CaughtUpFailureKind.barrier,
    code: code,
    message: message,
    targetCheckpoint: target,
    localCheckpoint: local,
  );
}

CaughtUpFailure? _checkpointFailure(String checkpoint, {required bool local}) {
  if (_canonicalCheckpoint.hasMatch(checkpoint)) return null;
  return _barrierFailure(
    local
        ? CaughtUpBarrierErrorCode.invalidLocalCheckpoint
        : CaughtUpBarrierErrorCode.invalidTarget,
    '${local ? 'local' : 'target'} checkpoint must be a canonical unsigned decimal string',
    target: local ? null : checkpoint,
    local: local ? checkpoint : null,
  );
}

CaughtUpOutcome<bool> checkpointReached(String local, String target) {
  final localError = _checkpointFailure(local, local: true);
  if (localError != null) return CaughtUpFailed<bool>(localError);
  final targetError = _checkpointFailure(target, local: false);
  if (targetError != null) return CaughtUpFailed<bool>(targetError);
  return CaughtUpSuccess<bool>(BigInt.parse(local) >= BigInt.parse(target));
}

CaughtUpFailure? validateAuthoritativeCheckpointTarget(
  AuthoritativeCheckpointTarget target, {
  String? expectedGeneration,
}) {
  if (target.protocolVersion != 1) {
    return _barrierFailure(
      CaughtUpBarrierErrorCode.invalidTarget,
      'authoritative checkpoint target has unsupported protocolVersion',
      target: target.checkpoint,
    );
  }
  final checkpointError = _checkpointFailure(target.checkpoint, local: false);
  if (checkpointError != null) return checkpointError;
  final generation = target.generation;
  if (generation != null && generation.isEmpty) {
    return _barrierFailure(
      CaughtUpBarrierErrorCode.invalidTarget,
      'authoritative checkpoint generation must be non-empty when present',
      target: target.checkpoint,
    );
  }
  if (expectedGeneration != null && generation != expectedGeneration) {
    return _barrierFailure(
      CaughtUpBarrierErrorCode.invalidated,
      'checkpoint generation does not match the expected generation',
      target: target.checkpoint,
    );
  }
  return null;
}

Duration _normalizedDuration(Duration value, Duration fallback) {
  if (value.isNegative) return Duration.zero;
  return value;
}

sealed class _WaitResult<T> {
  const _WaitResult();
}

final class _WaitValue<T> extends _WaitResult<T> {
  final T value;
  const _WaitValue(this.value);
}

final class _WaitOperationError<T> extends _WaitResult<T> {
  final Object cause;
  const _WaitOperationError(this.cause);
}

final class _WaitTimeout<T> extends _WaitResult<T> {
  const _WaitTimeout();
}

final class _WaitCancelled<T> extends _WaitResult<T> {
  const _WaitCancelled();
}

Future<_WaitResult<T>> _boundedForCaller<T>(
  Future<T> operation, {
  required Duration remaining,
  CaughtUpCancellationToken? cancellation,
}) {
  if (cancellation?.isCancelled ?? false) {
    return Future<_WaitResult<T>>.value(const _WaitCancelled<T>());
  }
  if (remaining <= Duration.zero) {
    return Future<_WaitResult<T>>.value(const _WaitTimeout<T>());
  }
  final operationResult = operation.then<_WaitResult<T>>(
    _WaitValue<T>.new,
    onError: (Object error) => _WaitOperationError<T>(error),
  );
  final timeoutResult = Future<_WaitResult<T>>.delayed(
    remaining,
    () => const _WaitTimeout<T>(),
  );
  if (cancellation == null) {
    return Future.any<_WaitResult<T>>(<Future<_WaitResult<T>>>[
      operationResult,
      timeoutResult,
    ]);
  }
  final cancellationResult = cancellation.whenCancelled.then<_WaitResult<T>>(
    (_) => const _WaitCancelled<T>(),
  );
  return Future.any<_WaitResult<T>>(<Future<_WaitResult<T>>>[
    operationResult,
    timeoutResult,
    cancellationResult,
  ]);
}

CaughtUpFailed<T> _waitFailure<T>(
  _WaitResult<T> result,
  String target,
  String local,
  String syncMessage,
) {
  if (result is _WaitOperationError<T>) {
    return CaughtUpFailed<T>(
      CaughtUpFailure(
        kind: CaughtUpFailureKind.sync,
        message: syncMessage,
        targetCheckpoint: target,
        localCheckpoint: local,
        cause: result.cause,
      ),
    );
  }
  if (result is _WaitCancelled<T>) {
    return CaughtUpFailed<T>(
      _barrierFailure(
        CaughtUpBarrierErrorCode.cancelled,
        'caught-up wait was cancelled',
        target: target,
        local: local,
      ),
    );
  }
  return CaughtUpFailed<T>(
    _barrierFailure(
      CaughtUpBarrierErrorCode.timeout,
      'caught-up wait timed out before the durable checkpoint reached the target',
      target: target,
      local: local,
    ),
  );
}

Future<CaughtUpOutcome<CaughtUpResult>> awaitCaughtUp(
  ProtocolSyncLoop loop,
  ProtocolQueueAdapter queue,
  AuthoritativeCheckpointTarget target, {
  Duration timeout = const Duration(seconds: 30),
  Duration pollInterval = const Duration(milliseconds: 50),
  CaughtUpCancellationToken? cancellation,
  String? expectedGeneration,
  bool Function()? isOnline,
  DateTime Function()? now,
}) async {
  final targetError = validateAuthoritativeCheckpointTarget(
    target,
    expectedGeneration: expectedGeneration,
  );
  if (targetError != null) return CaughtUpFailed<CaughtUpResult>(targetError);
  final clock = now ?? DateTime.now;
  final boundedTimeout = _normalizedDuration(timeout, const Duration(seconds: 30));
  final boundedPoll = _normalizedDuration(
    pollInterval,
    const Duration(milliseconds: 50),
  );
  final online = isOnline ?? loop.isOnline;
  final startedAt = clock();
  final deadline = startedAt.add(boundedTimeout);
  var checkpoint = await queue.pullCheckpoint();
  final initialCheckpointError = _checkpointFailure(checkpoint, local: true);
  if (initialCheckpointError != null) {
    return CaughtUpFailed<CaughtUpResult>(initialCheckpointError);
  }
  final initialReached = checkpointReached(checkpoint, target.checkpoint);
  if (initialReached is CaughtUpFailed<bool>) {
    return CaughtUpFailed<CaughtUpResult>(initialReached.error);
  }
  if ((initialReached as CaughtUpSuccess<bool>).value) {
    return CaughtUpSuccess<CaughtUpResult>(
      CaughtUpResult(
        targetCheckpoint: target.checkpoint,
        checkpoint: checkpoint,
        generation: target.generation,
        elapsed: clock().difference(startedAt),
        cycles: 0,
        alreadyCaughtUp: true,
      ),
    );
  }

  var cycles = 0;
  while (true) {
    if (cancellation?.isCancelled ?? false) {
      return CaughtUpFailed<CaughtUpResult>(
        _barrierFailure(
          CaughtUpBarrierErrorCode.cancelled,
          'caught-up wait was cancelled',
          target: target.checkpoint,
          local: checkpoint,
        ),
      );
    }
    if (!online() || loop.state.status == ProtocolSyncStatus.offline) {
      return CaughtUpFailed<CaughtUpResult>(
        _barrierFailure(
          CaughtUpBarrierErrorCode.offline,
          'cannot establish authoritative freshness while offline',
          target: target.checkpoint,
          local: checkpoint,
        ),
      );
    }
    final remaining = deadline.difference(clock());
    if (remaining <= Duration.zero) {
      return CaughtUpFailed<CaughtUpResult>(
        _barrierFailure(
          CaughtUpBarrierErrorCode.timeout,
          'caught-up wait timed out before the durable checkpoint reached the target',
          target: target.checkpoint,
          local: checkpoint,
        ),
      );
    }

    final before = checkpoint;
    final waited = await _boundedForCaller<ProtocolSyncCycleResult>(
      loop.syncNow(),
      remaining: remaining,
      cancellation: cancellation,
    );
    if (waited is! _WaitValue<ProtocolSyncCycleResult>) {
      return _waitFailure<ProtocolSyncCycleResult>(
        waited,
        target.checkpoint,
        checkpoint,
        'protocol sync failed before the caught-up target was reached',
      ) as CaughtUpFailed<CaughtUpResult>;
    }
    cycles++;

    checkpoint = await queue.pullCheckpoint();
    final checkpointError = _checkpointFailure(checkpoint, local: true);
    if (checkpointError != null) {
      return CaughtUpFailed<CaughtUpResult>(checkpointError);
    }
    final reached = checkpointReached(checkpoint, target.checkpoint);
    if (reached is CaughtUpFailed<bool>) {
      return CaughtUpFailed<CaughtUpResult>(reached.error);
    }
    if ((reached as CaughtUpSuccess<bool>).value) {
      return CaughtUpSuccess<CaughtUpResult>(
        CaughtUpResult(
          targetCheckpoint: target.checkpoint,
          checkpoint: checkpoint,
          generation: target.generation,
          elapsed: clock().difference(startedAt),
          cycles: cycles,
          alreadyCaughtUp: false,
        ),
      );
    }

    if (checkpoint == before && boundedPoll > Duration.zero) {
      final afterCycleRemaining = deadline.difference(clock());
      final delay = boundedPoll < afterCycleRemaining
          ? boundedPoll
          : afterCycleRemaining;
      final delayed = await _boundedForCaller<void>(
        Future<void>.delayed(delay),
        remaining: afterCycleRemaining,
        cancellation: cancellation,
      );
      if (delayed is! _WaitValue<void>) {
        final failure = _waitFailure<void>(
          delayed,
          target.checkpoint,
          checkpoint,
          'caught-up poll delay failed',
        );
        return CaughtUpFailed<CaughtUpResult>(failure.error);
      }
    }
  }
}

Future<CaughtUpOutcome<CaughtUpResult>> requestAndAwaitCaughtUp(
  ProtocolSyncLoop loop,
  ProtocolQueueAdapter queue,
  AuthoritativeCheckpointRequester requester, {
  Duration timeout = const Duration(seconds: 30),
  Duration pollInterval = const Duration(milliseconds: 50),
  CaughtUpCancellationToken? cancellation,
  String? expectedGeneration,
  bool Function()? isOnline,
  DateTime Function()? now,
}) async {
  final clock = now ?? DateTime.now;
  final boundedTimeout = _normalizedDuration(timeout, const Duration(seconds: 30));
  final startedAt = clock();
  final requestCancellation = CaughtUpCancellationToken();
  final mirror = cancellation == null
      ? null
      : Stream<void>.fromFuture(cancellation.whenCancelled).listen(
          (_) => requestCancellation.cancel(),
        );
  if (cancellation?.isCancelled ?? false) requestCancellation.cancel();
  final requested = await _boundedForCaller<AuthoritativeCheckpointTarget>(
    requester.requestCheckpoint(requestCancellation),
    remaining: boundedTimeout,
    cancellation: cancellation,
  );
  await mirror?.cancel();
  if (requested is _WaitOperationError<AuthoritativeCheckpointTarget>) {
    return CaughtUpFailed<CaughtUpResult>(
      CaughtUpFailure(
        kind: CaughtUpFailureKind.request,
        message: 'authoritative checkpoint request failed',
        cause: requested.cause,
      ),
    );
  }
  if (requested is _WaitCancelled<AuthoritativeCheckpointTarget>) {
    requestCancellation.cancel();
    return CaughtUpFailed<CaughtUpResult>(
      _barrierFailure(
        CaughtUpBarrierErrorCode.cancelled,
        'caught-up wait was cancelled',
      ),
    );
  }
  if (requested is _WaitTimeout<AuthoritativeCheckpointTarget>) {
    requestCancellation.cancel();
    return CaughtUpFailed<CaughtUpResult>(
      _barrierFailure(
        CaughtUpBarrierErrorCode.timeout,
        'caught-up wait timed out while requesting the authoritative checkpoint',
      ),
    );
  }
  final target = (requested as _WaitValue<AuthoritativeCheckpointTarget>).value;
  final elapsed = clock().difference(startedAt);
  final remaining = boundedTimeout - elapsed;
  return await awaitCaughtUp(
    loop,
    queue,
    target,
    timeout: remaining.isNegative ? Duration.zero : remaining,
    pollInterval: boundedPollInterval(pollInterval),
    cancellation: cancellation,
    expectedGeneration: expectedGeneration,
    isOnline: isOnline,
    now: clock,
  );
}

Duration boundedPollInterval(Duration pollInterval) {
  return _normalizedDuration(pollInterval, const Duration(milliseconds: 50));
}
